### ISSUE TITLE
How to adjust the dataset sampling method to probabilistic sampling instead of repeated sampling during training?

### ISSUE THREAD

**Issue by Monda-Hou** [AnjieCheng/NaVILA#41]
Hi, authors! Thanks for your good work!

I'm trying to add different datasets during training. I have already trained one  version which is r2r:rxr:human:scanqa=1:1:1:1. This version did not show well in area of action prediction. 

So I want to train one version which includes the Envdrop datasets which is good for action prediction. And I want to adjust the proportion of datasets, like r2r:rxr:envdrop:human=0.3:0.3:0.3:0.1. However, after checking the codes, it supports the form of r2r*3:rxr*3:envdrop*3:human*1. It means, only supports integer multiples(repeated sampling), not demical(probabilistic sampling).

So I have questions as follows:
**1. I'm not sure about whether my opinion(only supports integer multiples(repeated sampling), not demical(probabilistic sampling)) is right.
2. If only supports integer multiples, how to adjust it to probabilistic sampling, which will save the computing power and much time.** 
Could you please answer these questions? And if anyone have interests about this, please feel free to discuss :)

Have a good day!

**Comment by maxfanxd** [comment]
Same problem, I tried my dataset created by myself but get bad results. So I tried mixing it with r2r dataset with 0.25/0.25 proportion, but the result is also bad. I wonder if my dataset mixture method is wrong so I just used 0.2 percent of r2r dataset to train, but getting wrong output when evaluating. Have you handled this problem?

**Comment by Monda-Hou** [comment]
> Same problem, I tried my dataset created by myself but get bad results. So I tried mixing it with r2r dataset with 0.25/0.25 proportion, but the result is also bad. I wonder if my dataset mixture method is wrong so I just used 0.2 percent of r2r dataset to train, but getting wrong output when evaluating. Have you handled this problem?

Hi! What do you mean about bad results? I have solved this problem but not yet evaluated the model. I also commented your issue this morning. Do you want to have a conversation about this program?

**Comment by return-sleep** [comment]
Hi! Glad to hear that you’ve solved the dataset sampling issue! Could you please share a bit more detail on how you implemented probabilistic sampling (rather than integer-multiple repeated sampling)? Also, I noticed that NavLIA mentions a label rebalance strategy — may I ask if you have implemented that part as well? I found that in my own dataset, the model tends to over-predict a single action, and I wonder if label rebalancing could mitigate that. Thanks again for your help and for the nice discussion!

**Comment by Monda-Hou** [comment]
> Hi! Glad to hear that you’ve solved the dataset sampling issue! Could you please share a bit more detail on how you implemented probabilistic sampling (rather than integer-multiple repeated sampling)? Also, I noticed that NavLIA mentions a label rebalance strategy — may I ask if you have implemented that part as well? I found that in my own dataset, the model tends to over-predict a single action, and I wonder if label rebalancing could mitigate that. Thanks again for your help and for the nice discussion!

Hi! As for the sampling issue, I changed the code file(~/NaVILA/llava/data/builder.py) and added the probabilistic sampling in this file. You could check it. 


import os
from typing import Any, Optional

from hydra.utils import instantiate
from torch.utils.data import ConcatDataset, Dataset
from transformers import PreTrainedTokenizer

from llava.data.datasets_mixture import DATASETS_LEGACY
from llava.train.args import DataArguments, TrainingArguments
from llava.utils import io
from llava.utils.logging import logger
import random


__all__ = ["DATASETS", "register_datasets", "build_dataset"]


def register_datasets(name: Optional[str] = None):
    if name is None:
        name = os.environ.get("VILA_DATASETS", "default")
        logger.info(f"Registering datasets from `{name}`.")
    return io.load(os.path.join(os.path.dirname(__file__), "registry", f"{name}.yaml"))

DATASETS = register_datasets()

class RepeatedDataset(Dataset):
    def __init__(self, dataset: Dataset, times: int) -> None:
        super().__init__()
        self.dataset = dataset
        self.times = times

    def __len__(self) -> int:
        return len(self.dataset) * self.times

    def __getitem__(self, index: int) -> Any:
        return self.dataset[index % len(self.dataset)]

class WeightedMixtureDataset(Dataset):
    """随机按权重在多个子数据集里采样"""
    def __init__(self, datasets, probs):
        super().__init__()
        self.datasets = datasets
        self.probs = probs

        try:
            self._len = sum(len(d) for d in datasets)
        except Exception:
            self._len = 10**9  # fallback

    def __len__(self):
        return self._len

    def __getitem__(self, idx):
        ds = random.choices(self.datasets, weights=self.probs, k=1)[0]
        sample_idx = random.randrange(len(ds))
        return ds[sample_idx]


def build_dataset(
    mixture: str,
    data_args: DataArguments,
    training_args: TrainingArguments,
    tokenizer: PreTrainedTokenizer,
) -> Dataset:
    datasets = []
    probs = []

    for name in mixture.strip().lower().split("+"):
        weight = 1.0
        times = 1

        # 支持 name:prob
        if ":" in name:
            name, w = name.split(":")
            weight = float(w)

        # 支持 name*N
        if "*" in name:
            name, times = name.split("*")
            times = int(times)

        if name in DATASETS_LEGACY:
            logger.warning(f"Dataset {name} is registered under legacy mode.")
            dataset = build_dataset_legacy(
                name,
                data_args=data_args,
                training_args=training_args,
                tokenizer=tokenizer,
            )
        else:
            raise ValueError(f"Dataset {name} is not registered.")

        if times > 1:
            dataset = RepeatedDataset(dataset, times)

        datasets.append(dataset)
        probs.append(weight)

    # 归一化权重
    total = sum(probs)
    probs = [p / total for p in probs]

    logger.info(
        "Weighted mixture: " + ", ".join(f"{n}:{p:.3f}" for n, p in zip(mixture.split("+"), probs))
    )

    return WeightedMixtureDataset(datasets, probs)



def build_dataset_legacy(
    name: str,
    data_args: DataArguments,
    training_args: TrainingArguments,
    tokenizer: PreTrainedTokenizer,
) -> Dataset:
    from llava.data.dataset import DummyDataset, LazyEnvDropDataset, LazySupervisedDataset, LazyVLNCEDataset

    dataset = DATASETS_LEGACY[name]
    dataset_type = dataset.dataset_type
    if dataset_type == "torch":
        dataset_cls = LazySupervisedDataset
    elif dataset_type == "envdrop":
        dataset_cls = LazyEnvDropDataset
    elif dataset_type == "vlnce":
        dataset_cls = LazyVLNCEDataset
    else:
        raise NotImplementedError(f"{dataset_type} is not supported.")

    data_args.meta_path = getattr(dataset, "meta_path", None)
    data_args.caption_choice = getattr(dataset, "caption_choice", None)
    data_args.caption_choice_2 = getattr(dataset, "caption_choice_2", None)
    data_args.start_idx = getattr(dataset, "start_idx", None)
    data_args.end_idx = getattr(dataset, "end_idx", None)

    return dataset_cls(
        tokenizer=tokenizer,
        data_path=dataset.data_path,
        image_folder=getattr(dataset, "image_path"),
        data_args=data_args,
        training_args=training_args,
    )


There 's a little issue with this code which will print the result like "r2r:0.25:0.250". But it doesn't matter, it won't affect the training, just the print repeat. 

Btw, the description in sft_8frames.sh should be like "DATA_MIX="r2r:0.25+rxr:0.35+human:0.2+scanqa:0.05+envdrop:0.15" "

As for the second problem, sorry, I haven't noticed this problem. If you want, we could have a talk on this problem.
